samples/DispatcherSample.Web/MyCorsMiddleware.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace DispatcherSample.Web
+{
+    public class MyCorsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ICorsService _corsService;
+        private readonly ICorsPolicyProvider _corsPolicyProvider;
+        private readonly IOptions<CorsOptions> _options;
+
+        public MyCorsMiddleware(
+            RequestDelegate next,
+            ICorsService corsService,
+            ICorsPolicyProvider corsPolicyProvider,
+            IOptions<CorsOptions> options)
+        {
+            _next = next;
+            _corsService = corsService;
+            _corsPolicyProvider = corsPolicyProvider;
+            _options = options;
+        }
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            if (IsCorsEnabled(httpContext))
+            {
+                return InvokeInternal(httpContext);
+            }
+
+            return _next(httpContext);
+        }
+
+        private async Task InvokeInternal(HttpContext httpContext)
+        {
+            var corsPolicy = await GetCorsPolicyAsync(httpContext);
+            var corsResult = _corsService.EvaluatePolicy(httpContext, corsPolicy);
+            _corsService.ApplyResult(corsResult, httpContext.Response);
+
+            var accessControlRequestMethod = httpContext.Request.Headers[HeaderNames.AccessControlRequestMethod];
+            if (string.Equals(
+                    httpContext.Request.Method,
+                    HttpMethods.Options,
+                    StringComparison.OrdinalIgnoreCase) &&
+                    !StringValues.IsNullOrEmpty(accessControlRequestMethod))
+            {
+                // Since there is a policy which was identified,
+                // always respond to preflight requests.
+                httpContext.Response.StatusCode = StatusCodes.Status204NoContent;
+                return;
+            }
+
+            await _next(httpContext);
+        }
+
+        private bool IsCorsEnabled(HttpContext httpContext)
```

Comment:
You could make this into a `TryGetCorsPolicy` method and simplify things a bit further. The code as it is does the lookup twice.